src/lazy/encoder/binary/v1_1/fixed_uint.rs

```diff
@@ -0,0 +1,199 @@
+use std::io::Write;
+
+use num_bigint::BigUint;
+
+use crate::result::IonFailure;
+use crate::types::integer::UIntData;
+use crate::{IonResult, UInt};
+
+/// An Ion 1.1 encoding primitive that represents a fixed-length unsigned integer.
+#[derive(Debug)]
+pub struct FixedUInt {
+    value: UInt,
+    size_in_bytes: usize,
+}
+
+impl FixedUInt {
+    pub fn new(size_in_bytes: usize, value: impl Into<UInt>) -> Self {
+        Self {
+            value: value.into(),
+            size_in_bytes,
+        }
+    }
+
+    /// Reads a [`FixedUInt`] from the buffer.
+    ///
+    /// `input` is the byte slice from which to read a [`FixedUInt`].
+    /// `size_in_bytes` is the number of bytes to interpret as an unsigned integer.
+    /// `offset` is the position of the slice in some larger input stream. It is only used to populate
+    ///          an appropriate error message if reading fails.
+    #[inline]
+    pub fn read(input: &[u8], size_in_bytes: usize, offset: usize) -> IonResult<FixedUInt> {
+        if input.len() < size_in_bytes {
+            return IonResult::incomplete("reading a FixedUInt", offset);
+        }
+
+        let value: UInt = if input.len() <= 8 {
+            let mut buffer = [0; 8];
+            buffer[..size_in_bytes].copy_from_slice(input);
+            u64::from_le_bytes(buffer).into()
+        } else {
+            BigUint::from_bytes_le(&input[..size_in_bytes]).into()
+        };
+
+        Ok(FixedUInt::new(size_in_bytes, value))
+    }
+
+    #[inline]
+    fn write_u64<W: Write>(output: &mut W, value: u64) -> IonResult<usize> {
+        let encoded_bytes = value.to_le_bytes();
+        let leading_zeros = value.leading_zeros();
+        let num_encoded_bytes = (8 - (leading_zeros as usize / 8)).max(1);
```

Comment:
I don't know if `max` requires any branching, but if it does, I believe you could replace this equation with `(63 - leading_zeros) / 8 + 1`.

Reply:
`max` ends up using a branchless `cmov` ("conditional move") instruction.